src/mlpack/core/hpt/hpt.hpp

```diff
@@ -0,0 +1,268 @@
+   *   description for more information).
+   */
+  template<typename... Args>
+  TupleOfHyperParameters<Args...> Optimize(const Args&... args);
+
+  //! Access the performance measurement of the best model from the last run.
+  double BestObjective() const { return bestObjective; }
+
+  //! Access and modify the best model from the last run.
+  MLAlgorithm& BestModel() { return bestModel; }
+
+ private:
+  static_assert(
+      std::is_same<Optimizer, mlpack::optimization::GridSearch>::value,
+      "HyperParameterTuner now supports only the GridSearch optimizer");
+
+  /**
+   * A decorator that returns negated values of the original metric.
+   */
+  template<typename OriginalMetric>
+  struct Negated
+  {
+    static double Evaluate(MLAlgorithm& model,
+                           const MatType& xs,
+                           const PredictionsType& ys)
+    { return -OriginalMetric::Evaluate(model, xs, ys); }
+  };
+
+  //! A short alias for the full type of the cross-validation.
+  using CVType = typename std::conditional<Metric::NeedsMinimization,
+      CV<MLAlgorithm, Metric, MatType, PredictionsType, WeightsType>,
+      CV<MLAlgorithm, Negated<Metric>, MatType, PredictionsType,
+          WeightsType>>::type;
+
+
+  //! The cross-validation object for assessing sets of hyper-parameters.
+  CVType cv;
+
+  //! The optimizer.
+  Optimizer optimizer;
+
+  //! The best objective from the last run.
+  double bestObjective;
+
+  //! The best model from the last run.
+  MLAlgorithm bestModel;
+
+  /**
+   * The set of methods to initialize the GridSearh optimizer. We basically need
+   * to go through all arguments passed to the Optimize method, gather all
+   * non-bound arguments (collections) and pass them into the GridSearch
+   * constructor.
+   */
+  template<size_t I,
+           typename ArgsTuple,
+           typename... Collections,
+           typename = std::enable_if_t<I == std::tuple_size<ArgsTuple>::value>>
+  inline void InitGridSearch(const ArgsTuple& args,
+                             Collections... collections);
+
+  template<size_t I,
+           typename ArgsTuple,
+           typename... Collections,
+           typename = std::enable_if_t<I < std::tuple_size<ArgsTuple>::value>,
+           typename = std::enable_if_t<IsPreBoundArg<
+               typename std::tuple_element<I, ArgsTuple>::type>::value>>
+  inline void InitGridSearch(const ArgsTuple& args,
+                             Collections... collections);
+
+  template<size_t I,
+           typename ArgsTuple,
+           typename... Collections,
+           typename = std::enable_if_t<I < std::tuple_size<ArgsTuple>::value>,
+           typename = std::enable_if_t<!IsPreBoundArg<
+               typename std::tuple_element<I, ArgsTuple>::type>::value>,
+           typename = void>
+  inline void InitGridSearch(const ArgsTuple& args,
+                             Collections... collections);
```

Comment:
Return-type SFINAE doesn't help much since we can't put everything here into one condition (if we do compilers dosn't compile it). So, if we don't want explicitly specify variadic templates, I guess it's better to stay it as it is.